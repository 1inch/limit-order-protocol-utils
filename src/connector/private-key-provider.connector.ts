import {ProviderConnector} from './provider.connector';
import Web3 from 'web3';
import {EIP712TypedData} from '../model/eip712.model';
import {AbiItem} from '../model/abi.model';
import {AbiItem as Web3AbiItem} from 'web3-utils';
import {MsgParams, signTypedData_v4} from 'eth-sig-util';

export class PrivateKeyProviderConnector implements ProviderConnector {
    constructor(
        private readonly privateKey: string,
        protected readonly web3Provider: Web3
    ) {}

    contractEncodeABI(
        abi: AbiItem[],
        address: string | null,
        methodName: string,
        methodParams: unknown[]
    ): string {
        const contract = new this.web3Provider.eth.Contract(
            abi as Web3AbiItem[],
            address === null ? undefined : address
        );

        return contract.methods[methodName](...methodParams).encodeABI();
    }

    signTypedData(
        _walletAddress: string,
        typedData: EIP712TypedData,
        /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
        _typedDataHash = ''
    ): Promise<string> {
        const result = signTypedData_v4(Buffer.from(this.privateKey, 'hex'), {
            data: typedData,
            /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
        } as MsgParams<any>);

        return Promise.resolve(result);
    }

    ethCall(contractAddress: string, callData: string): Promise<string> {
        return this.web3Provider.eth.call({
            to: contractAddress,
            data: callData,
        });
    }

    decodeABIParameter<T>(type: string, hex: string): T {
        return this.web3Provider.eth.abi.decodeParameter(type, hex) as T;
    }
}
