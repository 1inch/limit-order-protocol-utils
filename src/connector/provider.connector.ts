import {EIP712TypedData} from '../model/eip712.model';
import {AbiItem, AbiOutput} from '../model/abi.model';

export interface AbiDecodeResult {
    [key: string]: unknown,
}

export interface ProviderConnector {
    contractEncodeABI(
        abi: AbiItem[],
        address: string | null,
        methodName: string,
        methodParams: unknown[]
    ): string;

    signTypedData(
        walletAddress: string,
        typedData: EIP712TypedData,
        typedDataHash: string
    ): Promise<string>;

    ethCall(contractAddress: string, callData: string): Promise<string>;

    decodeABIParameter<T>(type: AbiOutput | string, hex: string): T;

    decodeABICallParameters(types: Array<AbiOutput | string>, callData: string): AbiDecodeResult;
}
