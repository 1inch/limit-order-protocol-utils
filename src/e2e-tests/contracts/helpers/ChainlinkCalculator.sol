// SPDX-License-Identifier: MIT

pragma solidity 0.8.19;

import "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
import "@openzeppelin/contracts/utils/math/SafeCast.sol";


/// @title A helper contract for interactions with https://docs.chain.link
contract ChainlinkCalculator {
    using SafeCast for int256;

    error DifferentOracleDecimals();

    uint256 private constant _SPREAD_DENOMINATOR = 1e9;
    uint256 private constant _INVERSE_MASK = 1 << 255;

    /// @notice Calculates price of token relative to oracle unit (ETH or USD)
    /// @param inverseAndSpread concatenated inverse flag and spread.
    /// Lowest 254 bits specify spread amount. Spread is scaled by 1e9, i.e. 101% = 1.01e9, 99% = 0.99e9.
    /// Highest bit is set when oracle price should be inverted,
    /// e.g. for DAI-ETH oracle, inverse=false means that we request DAI price in ETH
    /// and inverse=true means that we request ETH price in DAI
    /// @return Amount * spread * oracle price
    function singlePrice(AggregatorV3Interface oracle, uint256 inverseAndSpread, uint256 amount) external view returns(uint256) {
        (, int256 latestAnswer,,,) = oracle.latestRoundData();
        bool inverse = inverseAndSpread & _INVERSE_MASK > 0;
        uint256 spread = inverseAndSpread & (~_INVERSE_MASK);
        if (inverse) {
            return amount * spread * (10 ** oracle.decimals()) / latestAnswer.toUint256() / _SPREAD_DENOMINATOR;
        } else {
            return amount * spread * latestAnswer.toUint256() / (10 ** oracle.decimals()) / _SPREAD_DENOMINATOR;
        }
    }

    /// @notice Calculates price of token A relative to token B. Note that order is important
    /// @return result Token A relative price times amount
    function doublePrice(AggregatorV3Interface oracle1, AggregatorV3Interface oracle2, uint256 spread, int256 decimalsScale, uint256 amount) external view returns(uint256 result) {
        if (oracle1.decimals() != oracle2.decimals()) revert DifferentOracleDecimals();

        (, int256 latestAnswer1,,,) = oracle1.latestRoundData();
        result = amount * spread * latestAnswer1.toUint256();

        if (decimalsScale > 0) {
            result *= 10 ** decimalsScale.toUint256();
        } else if (decimalsScale < 0) {
            result /= 10 ** (-decimalsScale).toUint256();
        }

        (, int256 latestAnswer2,,,) = oracle2.latestRoundData();
        result /= latestAnswer2.toUint256() * _SPREAD_DENOMINATOR;
    }
}
