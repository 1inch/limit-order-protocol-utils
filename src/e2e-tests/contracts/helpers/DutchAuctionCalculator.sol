// SPDX-License-Identifier: MIT

pragma solidity 0.8.19;

import "@openzeppelin/contracts/utils/math/Math.sol";

contract DutchAuctionCalculator {
    uint256 private constant _LOW_128_BITS = 0xffffffffffffffffffffffffffffffff;

    function getMakingAmount(uint256 startTimeEndTime, uint256 takingAmountStart, uint256 takingAmountEnd, uint256 makingAmount, uint256 requestedTakingAmount) external view returns(uint256) {
        uint256 calculatedTakingAmount = _calculateAuctionTakingAmount(startTimeEndTime, takingAmountStart, takingAmountEnd);
        return requestedTakingAmount * makingAmount / calculatedTakingAmount;
    }

    function getTakingAmount(uint256 startTimeEndTime, uint256 takingAmountStart, uint256 takingAmountEnd, uint256 makingAmount, uint256 requestedMakingAmount) external view returns(uint256) {
        uint256 calculatedTakingAmount = _calculateAuctionTakingAmount(startTimeEndTime, takingAmountStart, takingAmountEnd);
        return (requestedMakingAmount * calculatedTakingAmount + makingAmount - 1) / makingAmount;
    }

    function _calculateAuctionTakingAmount(uint256 startTimeEndTime, uint256 takingAmountStart, uint256 takingAmountEnd) private view returns(uint256) {
        uint256 startTime = startTimeEndTime >> 128;
        uint256 endTime = startTimeEndTime & _LOW_128_BITS;
        uint256 currentTime = Math.max(startTime, Math.min(endTime, block.timestamp));  // solhint-disable-line not-rely-on-time
        return (takingAmountStart * (endTime - currentTime) + takingAmountEnd * (currentTime - startTime)) / (endTime - startTime);
    }
}
