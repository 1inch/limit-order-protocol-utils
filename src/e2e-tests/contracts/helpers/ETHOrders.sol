// SPDX-License-Identifier: MIT

pragma solidity 0.8.19;

import { EIP712 } from "@openzeppelin/contracts/utils/cryptography/EIP712.sol";
import "@1inch/solidity-utils/contracts/libraries/SafeERC20.sol";
import "@1inch/solidity-utils/contracts/OnlyWethReceiver.sol";

import "../interfaces/IPostInteraction.sol";
import "../OrderLib.sol";

/// @title ETH limit orders contract
contract ETHOrders is IPostInteraction, OnlyWethReceiver {
    using SafeERC20 for IWETH;
    using OrderLib for IOrderMixin.Order;
    using MakerTraitsLib for MakerTraits;
    using ExtensionLib for bytes;
    using AddressLib for Address;

    error AccessDenied();
    error InvalidOrder();
    error NotEnoughBalance();
    error ExistingOrder();

    /// @notice ETH order struct.
    struct ETHOrder {
        address maker;
        uint96 balance;
    }

    address private immutable _limitOrderProtocol;
    IWETH private immutable _WETH; // solhint-disable-line var-name-mixedcase
    /// @notice Makers and their uint96 ETH balances in single mapping.
    mapping(bytes32 => ETHOrder) public ordersMakersBalances;

    event ETHDeposited(bytes32 orderHash, uint256 amount);
    event ETHOrderCancelled(bytes32 orderHash, uint256 amount);

    /// @notice Only limit order protocol can call this contract.
    modifier onlyLimitOrderProtocol() {
        if (msg.sender != _limitOrderProtocol) revert AccessDenied();

        _;
    }

    constructor(IWETH weth, address limitOrderProtocol) OnlyWethReceiver(address(weth)) {
        _WETH = weth;
        _limitOrderProtocol = limitOrderProtocol;
        _WETH.approve(limitOrderProtocol, type(uint256).max);
    }

    /*
     * @notice Returns batch of eth order for batch of orders hashes.
     */
    function ethOrdersBatch(bytes32[] calldata orderHashes) external view returns(ETHOrder[] memory ethOrders) {
        ethOrders = new ETHOrder[](orderHashes.length);
        for (uint256 i = 0; i < orderHashes.length; i++) {
            ethOrders[i] = ordersMakersBalances[orderHashes[i]];
        }
    }

    /*
     * @notice Checks if ETH order is valid, makes ETH deposit for an order, saves real maker and wraps ETH into WETH.
     */
    function ethOrderDeposit(IOrderMixin.Order calldata order, bytes calldata extension) external payable returns(bytes32 orderHash) {
        if (!order.makerTraits.needPostInteractionCall()) revert InvalidOrder();
        order.validateExtension(extension);
        if (order.maker.get() != address(this)) revert AccessDenied();
        if (order.getReceiver() != msg.sender) revert AccessDenied();
        if (order.makingAmount != msg.value) revert InvalidOrder();
        bytes calldata interaction = extension.postInteractionTargetAndData();
        if (interaction.length != 20 || address(bytes20(interaction)) != address(this)) revert InvalidOrder();
        orderHash = IOrderMixin(_limitOrderProtocol).hashOrder(order);
        if (ordersMakersBalances[orderHash].maker != address(0)) revert ExistingOrder();
        ordersMakersBalances[orderHash] = ETHOrder({
            maker: msg.sender,
            balance: uint96(msg.value)
        });
        _WETH.safeDeposit(msg.value);
        emit ETHDeposited(orderHash, msg.value);
    }

    /**
     * @notice Sets ordersMakersBalances to 0, refunds ETH and does standard order cancellation on Limit Order Protocol.
     */
    function cancelOrder(MakerTraits makerTraits, bytes32 orderHash) external {
        if (ordersMakersBalances[orderHash].maker != msg.sender) revert InvalidOrder();
        IOrderMixin(_limitOrderProtocol).cancelOrder(makerTraits, orderHash);
        uint256 refundETHAmount = ordersMakersBalances[orderHash].balance;
        ordersMakersBalances[orderHash].balance = 0;
        _WETH.safeWithdrawTo(refundETHAmount, msg.sender);
        emit ETHOrderCancelled(orderHash, refundETHAmount);
    }

    /**
     * @notice Checks if orderHash signature was signed with real order maker.
     */
    function isValidSignature(bytes32 orderHash, bytes calldata signature) external view returns(bytes4) {
        if (ECDSA.recoverOrIsValidSignature(ordersMakersBalances[orderHash].maker, orderHash, signature)) {
            return IERC1271.isValidSignature.selector;
        } else {
            return 0xffffffff;
        }
    }

    /**
     * @notice Callback method that gets called after all funds transfers.
     * Updates _ordersMakersBalances by makingAmount for order with orderHash.
     * @param orderHash Hash of the order being processed
     * @param makingAmount Actual making amount
     */
    function postInteraction(
        IOrderMixin.Order calldata /*order*/,
        bytes32 orderHash,
        address /*taker*/,
        uint256 makingAmount,
        uint256 /*takingAmount*/,
        uint256 /*remainingMakingAmount*/,
        bytes calldata /*extraData*/
    ) external onlyLimitOrderProtocol {
        if (ordersMakersBalances[orderHash].balance < makingAmount) revert NotEnoughBalance();
        unchecked {
            ordersMakersBalances[orderHash].balance -= uint96(makingAmount);
        }
    }
}
