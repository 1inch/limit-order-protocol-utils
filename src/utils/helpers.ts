/**
 * Determine whether the given `input` is iterable.
 *
 * @returns {Boolean}
 */
 export function isIterable(input: unknown): input is Iterable<unknown> {  
    if (input === null || typeof input !== 'object') {
      return false
    }

    if (input instanceof Array) return true;
  
    return typeof (input as Iterable<unknown>)[Symbol.iterator] === 'function'
}


export function mapObject<Index extends string, ValueBefore, ValueAfter>(
    obj: Record<Index, ValueBefore>,
    callbackfn: (value: ValueBefore, key: Index) => ValueAfter,
): {
    [key in Index]: ValueAfter
} {
    const result: Record<Index, ValueAfter> = {} as Record<Index, ValueAfter>;

    Object.entries(obj).forEach(
        ([key, value]) => {
            result[key as Index] = callbackfn(value as ValueBefore, key as Index);
        }
    );

    return result;
}

export function mapAny<ValueBefore, ValueAfter>(
    obj: Iterable<ValueBefore> | Record<string, ValueBefore>,
    callbackfn: (value: ValueBefore, key: string | number) => ValueAfter,
): ValueAfter[] | Record<string, ValueAfter>
{
    if (isIterable(obj)) {
        return Array.from(obj).map(callbackfn);
    }

    return mapObject(obj, callbackfn);
}