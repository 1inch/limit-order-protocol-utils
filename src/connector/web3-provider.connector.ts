import {ProviderConnector} from './provider.connector';
import Web3 from 'web3';
import {EIP712TypedData} from '../model/eip712.model';
import {AbiItem} from '../model/abi.model';
import {AbiItem as Web3AbiItem} from 'web3-utils';

export class Web3ProviderConnector implements ProviderConnector {
    constructor(protected readonly web3Provider: Web3) {}

    contractEncodeABI(
        abi: AbiItem[],
        address: string | null,
        methodName: string,
        methodParams: unknown[]
    ): string {
        const contract = new this.web3Provider.eth.Contract(
            abi as Web3AbiItem[],
            address === null ? undefined : address
        );

        return contract.methods[methodName](...methodParams).encodeABI();
    }

    signTypedData(
        walletAddress: string,
        typedData: EIP712TypedData,
        /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
        _typedDataHash: string
    ): Promise<string> {
        if (!this.web3Provider.currentProvider) {
            throw new Error('Web3 currentProvider is null');
        }

        return this.web3Provider.currentProvider.send('eth_signTypedData_v4', [
            walletAddress,
            JSON.stringify(typedData),
        ]);
    }

    ethCall(contractAddress: string, callData: string): Promise<string> {
        return this.web3Provider.eth.call({
            to: contractAddress,
            data: callData,
        });
    }

    decodeABIParameter<T>(type: string, hex: string): T {
        return this.web3Provider.eth.abi.decodeParameter(type, hex) as T;
    }
}
