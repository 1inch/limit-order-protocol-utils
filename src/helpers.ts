/**
 * Public utils interface
 */
import { ZX } from "./limit-order-protocol.const";
import { LimitOrder, LimitOrderInteractions } from "./model/limit-order-protocol.model";
import { joinStaticCalls, parseInteractionForField, trim0x } from "./utils/limit-order.utils";

export { trim0x } from './utils/limit-order.utils';


export const InteractionsFields = {
    makerAssetData: 0,
    takerAssetData: 1,
    getMakingAmount: 2,
    getTakingAmount: 3,
    predicate: 4,
    permit: 5,
    preInteraction: 6,
    postInteraction: 7,
// cuz enum has numeric keys also
} as const;

export type InteractionName = keyof typeof InteractionsFields;

export type Interactions = {
    [key in InteractionName]: string;
};

// eslint-disable-next-line max-lines-per-function
export function packInteractions({
    makerAssetData = ZX,
    takerAssetData = ZX,
    getMakingAmount = ZX,
    getTakingAmount = ZX,
    predicate = ZX,
    permit = ZX,
    preInteraction = ZX,
    postInteraction = ZX,
}: Partial<Interactions>): LimitOrderInteractions {
    const allInteractions = [
        makerAssetData,
        takerAssetData,
        getMakingAmount,
        getTakingAmount,
        predicate,
        permit,
        preInteraction,
        postInteraction,
    ];

    const { offsets, data: interactions } = joinStaticCalls(allInteractions);
    return { offsets, interactions };
}

export function unpackInteractions(offsets: string | bigint, interactions: string): Interactions {
    const offsetsBN = BigInt(offsets);

    const parsedInteractions = {} as Interactions;

    Object.entries(InteractionsFields).forEach(([name, position]) => {
        parsedInteractions[name as InteractionName] = parseInteractionForField(
            offsetsBN,
            interactions,
            position as number,
        );
    });

    return parsedInteractions
}

export function unpackInteraction<T extends InteractionName>(
    order: LimitOrder,
    name: T,
): Interactions[T] {
    return parseInteractionForField(
        BigInt(order.offsets),
        order.interactions,
        InteractionsFields[name],
    )
}

/**
 * @returns `true` if interaction value is empty of 0x
 */
export function hasInteraction(order: LimitOrder, name: InteractionName): boolean {
    const interaction = unpackInteraction(order, name);

    return trim0x(interaction) !== '';
}
