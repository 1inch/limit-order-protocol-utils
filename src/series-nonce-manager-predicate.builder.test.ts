
import { ChainId } from "./model/limit-order-protocol.model";
import {BETA_CONTRACT_ADDRESSES, mocksForChain, mocksForV3Chain} from "./test/helpers";
import { NonceSeriesV2 } from "./model/series-nonce-manager.model";
import { SeriesNonceManagerPredicateBuilder } from "./series-nonce-manager-predicate.builder";
import { ProviderConnector } from "./connector/provider.connector";


describe("SeriesNonceManagerFacade", () => {
    const walletAddress = '0x1c667c6308d6c9c8ce5bd207f524041f67dbc65e';
    let seriesNonceManagerContractAddress: string;

    let seriesNonceManagerPredicateBuilder: SeriesNonceManagerPredicateBuilder;
    let providerConnector: ProviderConnector;

    beforeEach(() => {
        const chainId = ChainId.etherumMainnet;

        const mocks = mocksForV3Chain(chainId, BETA_CONTRACT_ADDRESSES[chainId]);
        providerConnector = mocks.providerConnector;
        seriesNonceManagerPredicateBuilder = mocks.seriesNonceManagerPredicateBuilder;
        seriesNonceManagerContractAddress = mocks.seriesNonceManagerContractAddress;
    });

    it("nonce", () => {
        expect(
            seriesNonceManagerPredicateBuilder.nonce(NonceSeriesV2.P2PV3, walletAddress),
        ).toMatchSnapshot();
    });

    xdescribe("web3 calls", () => {
        it("nonceEquals call", async () => {
            const calldata = seriesNonceManagerPredicateBuilder.nonceEquals(NonceSeriesV2.LimitOrderV3, walletAddress, 4);

            const result = await providerConnector.ethCall(seriesNonceManagerContractAddress, calldata);
            expect(result).toBe('0x0000000000000000000000000000000000000000000000000000000000000001');
        });
    });
});
