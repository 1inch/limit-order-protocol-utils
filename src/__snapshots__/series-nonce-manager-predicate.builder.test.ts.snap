// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeriesNonceManagerFacade nonce 1`] = `"0x4a7f2a4b00000000000000000000000000000000000000000000000000000000000000010000000000000000000000001c667c6308d6c9c8ce5bd207f524041f67dbc65e"`;

exports[`SeriesNonceManagerFacade nonceEquals 1`] = `"0x9762222100000000000000000000000000000000000000000000000000000000000000010000000000000000000000001c667c6308d6c9c8ce5bd207f524041f67dbc65e0000000000000000000000000000000000000000000000000000000000000065"`;

exports[`SeriesNonceManagerFacade nonceEquals 2`] = `"0x2cc2878d0144838fc0000000006500011c667c6308d6c9c8ce5bd207f524041f67dbc65e"`;
